### PR TITLE
MAID-3233: add test for unpolled and unconsensused observations

### PR DESCRIPTION
**Pull request by Fraser999**
I'll add the WIP tag to this since I think it should be merged after #150 to avoid causing any more pain to that much larger PR.

It should be fairly easy to regenerate the dot file for this test once the dot-writing/parsing changes are in place.  This one was generated using the dot_gen tool with the following scenario:
```rust
let _ = scenarios
    .add(
        "parsec::functional_tests::unpolled_and_unconsensused_observations",
        |env| {
            let obs = ObservationSchedule {
                genesis: peer_ids!("Alice", "Bob", "Carol", "Dave"),
                schedule: vec![(0, AddPeer(PeerId::new("Eric")))],
            };

            Schedule::from_observation_schedule(env, &ScheduleOptions::default(), obs)
        },
    ).file("Alice", "alice.dot")
    .seed([1, 3, 1, 4]);
```

### CODE REVIEW

File: src/parsec.rs [hunk 1]
```diff
@@ -1802,19 +1802,26 @@ impl Parsec<Transaction, PeerId> {
...
+        for event in parsed_contents.events.values() {
+            if let Some(payload) = event.vote().map(Vote::payload) {
+                let observation = parsec
+                    .observations
+                    .entry(payload.create_hash())
+                    .or_insert_with(ObservationInfo::default);
+
+                if *event.creator() == our_pub_id {
+                    observation.created_by_us = true;
+                }
```
Comment (octol) [inline src/parsec.rs]: Help me understand why the two blocks were swapped around, is there a dependency I'm missing? Is the main difference the `created_by_us` variable being set? Was thinking maybe this change can be it's own commit that goes before the test?

Comment (Fraser999) [reply src/parsec.rs]: You're right - let me break this into two commits.  I'll explain in the commit message.

Comment (octol) [reply src/parsec.rs]: Nice, makes it easier to see why it was done :)

File: src/parsec.rs [hunk 2]
```diff
@@ -2123,7 +2130,7 @@ mod functional_tests {
 
         let mut alice = Parsec::from_parsed_contents(parsed_contents);
         initialise_membership_lists(&mut alice.peer_list);
-        let genesis_group: BTreeSet<_> = alice.peer_list.all_ids().into_iter().cloned().collect();
+        let genesis_group: BTreeSet<_> = alice.peer_list.all_ids().cloned().collect();
```
Comment (octol) [inline src/parsec.rs]: Of course personally I think this should also be it's own commit ;)
...but it's so tiny that probably not worth extracting out at this point

File: src/parsec.rs [hunk 3]
```diff
@@ -2607,4 +2614,69 @@ mod functional_tests {
...
+        let genesis = Observation::Genesis(mock::create_ids(4).into_iter().collect());
+        {
+            let mut unpolled_observations = alice.our_unpolled_observations();
+            assert_eq!(*unwrap!(unpolled_observations.next()), genesis);
+            assert_eq!(*unwrap!(unpolled_observations.next()), add_eric);
+            assert!(unpolled_observations.next().is_none());
+        }
+
+        // Call `poll()` and retry - should only return our vote for `Add(Eric)`.
+        unwrap!(alice.poll());
```
Comment (octol) [inline src/parsec.rs]: Possible to check the returned block from `poll`? Would make it slightly quicker to read what's happening. That said it's still very readable what's going on...

File: src/parsec.rs [hunk 4]
```diff
@@ -2607,4 +2614,69 @@ mod functional_tests {
...
+        unwrap!(alice.add_event(b_17));
+
+        // Since we haven't call `poll()` again yet, should still return our vote for `Add(Eric)`.
+        // However, `has_unconsensused_observations()` should now return false.
+        assert!(!alice.has_unconsensused_observations());
+        assert_eq!(alice.our_unpolled_observations().count(), 1);
+        assert_eq!(*unwrap!(alice.our_unpolled_observations().next()), add_eric);
+
+        // Call `poll()` and retry - should return none.
+        unwrap!(alice.poll());
```
Comment (octol) [inline src/parsec.rs]: Same as the other `poll` comment

File: src/parsec.rs (hunk 3 above)
Comment (Fraser999) [reply src/parsec.rs]: I did nearly do that actually, but decided against it in order to keep the test focused on what was relevant to the task; i.e. unpolled observations and our unconsensused observations.

File: src/parsec.rs [hunk 5]
```diff
@@ -2614,4 +2614,69 @@ mod functional_tests {
...
+
+    #[test]
+    fn unpolled_and_unconsensused_observations() {
+        let mut alice_contents = parse_test_dot_file("alice.dot");
+        let b_17 = unwrap!(alice_contents.remove_latest_event());
+
+        let mut alice = Parsec::from_parsed_contents(alice_contents);
+        alice.restart_consensus(); // This is needed so the Genesis observation is consensused.
+
+        // `Add(Eric)` should still be unconsensused.
```
Comment (pierrechevalier83) [inline src/parsec.rs]: I would add: `since b_17 would be the first gossip event to reach consensus on Add(Eric), but it was removed from the graph`.

File: src/parsec.rs [hunk 6]
```diff
@@ -2614,4 +2614,69 @@ mod functional_tests {
...
+            assert_eq!(*unwrap!(unpolled_observations.next()), genesis);
+            assert_eq!(*unwrap!(unpolled_observations.next()), add_eric);
+            assert!(unpolled_observations.next().is_none());
+        }
+
+        // Call `poll()` and retry - should only return our vote for `Add(Eric)`.
+        unwrap!(alice.poll());
+        assert!(alice.poll().is_none());
+        assert!(alice.has_unconsensused_observations());
+        assert_eq!(alice.our_unpolled_observations().count(), 1);
```
Comment (pierrechevalier83) [inline src/parsec.rs]: nitpick: instead of checking the size is 1 and then checking next, I would check next twice and make sure it's none the second time to be more symmetrical with the previous check. (Super minor but i like symmetry :p )

File: src/parsec.rs [hunk 7]
```diff
@@ -2614,4 +2614,69 @@ mod functional_tests {
...
+        assert_eq!(*unwrap!(alice.our_unpolled_observations().next()), add_eric);
+
+        // Have Alice process B_17 to get consensus on `Add(Eric)`.
+        unwrap!(alice.add_event(b_17));
+
+        // Since we haven't call `poll()` again yet, should still return our vote for `Add(Eric)`.
+        // However, `has_unconsensused_observations()` should now return false.
+        assert!(!alice.has_unconsensused_observations());
+        assert_eq!(alice.our_unpolled_observations().count(), 1);
```
Comment (pierrechevalier83) [inline src/parsec.rs]: same as above

File: src/parsec.rs (hunk 6 above)
Comment (Fraser999) [reply src/parsec.rs]: Same here, but I preferred this option to having to keep the iterator in a variable since that variable needs to be destroyed once we're finished the check (hence the extra level of nesting in the previous check)
